File: sources/Deltares.Probabilistic/Model/ModelSample.cpp
// Copyright (C) Stichting Deltares. All rights reserved.
//
// This file is part of the Probabilistic Library.
//
// The Probabilistic Library is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.
//
#include "ModelSample.h"

namespace Deltares::Models
{
    void ModelSample::clear()
    {
        IterationIndex = -1;
        threadId = 0;
        Weight = std::nan("");
        AllowProxy = true;
        UsedProxy = false;
        IsRestartRequired = false;
        Z = std::nan("");
        Beta = 0;
        Tag = 0;
    }

    /**
     * \brief Copies the results from another sample
     */
    void ModelSample::copyFrom(const std::shared_ptr<ModelSample>& source)
    {
        this->Z = source->Z;
        this->OutputValues = std::vector<double>(source->OutputValues.size());
        for (size_t i = 0; i < this->OutputValues.size(); i++)
        {
            this->OutputValues[i] = source->OutputValues[i];
        }
    }

    void ModelSample::fillModelSampleStruct(ModelSampleStruct* sampleStruct) const
    {
        sampleStruct->Values = Values.data();
        sampleStruct->ValuesCount = static_cast<int>(Values.size());

        sampleStruct->OutputValues = OutputValues.data();
        sampleStruct->OutputValuesCount = static_cast<int>(OutputValues.size());

        sampleStruct->IterationIndex = IterationIndex;
        sampleStruct->threadId = threadId;
        sampleStruct->Weight = Weight;
        sampleStruct->AllowProxy = AllowProxy;
        sampleStruct->UsedProxy = UsedProxy;
        sampleStruct->IsRestartRequired = IsRestartRequired;
        sampleStruct->Beta = Beta;
        sampleStruct->Z = Z;
        sampleStruct->ExtendedLogging = ExtendedLogging;
        sampleStruct->LoggingCounter = LoggingCounter;
        sampleStruct->Tag = Tag;
    }

    void ModelSample::setModelSampleStruct(ModelSampleStruct* sampleStruct)
    {
        Values.resize(sampleStruct->ValuesCount);
        for (int i = 0; i < sampleStruct->ValuesCount; i++)
        {
            Values[i] = sampleStruct->Values[i];
        }

        OutputValues.resize(sampleStruct->OutputValuesCount);
        for (int i = 0; i < sampleStruct->OutputValuesCount; i++)
        {
            OutputValues[i] = sampleStruct->OutputValues[i];
        }

        IterationIndex = sampleStruct->IterationIndex;
        threadId = sampleStruct->threadId;
        Weight = sampleStruct->Weight;
        AllowProxy = sampleStruct->AllowProxy;
        UsedProxy = sampleStruct->UsedProxy;
        IsRestartRequired = sampleStruct->IsRestartRequired;
        Beta = sampleStruct->Beta;
        Z = sampleStruct->Z;
        ExtendedLogging = sampleStruct->ExtendedLogging;
        LoggingCounter = sampleStruct->LoggingCounter;
        Tag = sampleStruct->Tag;
    }

    bool ModelSample::hasSameValues(std::shared_ptr<ModelSample> other)
    {
        if (this->Values.size() != other->Values.size())
        {
            return false;
        }

        for (int i = 0; i < this->Values.size(); i++)
        {
            if (this->Values[i] != other->Values[i])
            {
                return false;
            }
        }

        return true;
    }
}

File: sources/Deltares.Probabilistic.NetCS/Model/ModelSample.cs
// Copyright (C) Stichting Deltares. All rights reserved.
//
// This file is part of the Probabilistic Library.
//
// The Probabilistic Library is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.
//
using System;
using System.Linq;
using System.Runtime.InteropServices;

namespace Deltares.Probabilistic.Model;

[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
public delegate void ModelSampleCallback(ref ModelSampleStruct sample);

[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
public delegate void MultipleModelSampleCallback(IntPtr samples, int count);

public class ModelSample
{
    private double[] values = null;
    private double[] outputValues = null;

    public ModelSample()
    {
    }

    public ModelSample(double[] values)
    {
        // copy values
        this.values = values.Select(p => p).ToArray();
    }

    public ModelSample(ModelSampleStruct sample)
    {
        // Copy from unmanaged → managed
        values = new double[sample.ValuesCount];
        if (sample.ValuesCount > 0)
        {
            Marshal.Copy(sample.Values, values, 0, sample.ValuesCount);
        }

        outputValues = new double[sample.OutputValuesCount];
        if (sample.OutputValuesCount > 0)
        {
            Marshal.Copy(sample.OutputValues, outputValues, 0, sample.OutputValuesCount);
        }

        this.Iteration = sample.IterationIndex;
        this.Beta = sample.Beta;
        this.Z = sample.Z;
        this.Weight = sample.Weight;
        this.AllowProxy = sample.AllowProxy;
        this.UsedProxy = sample.UsedProxy;
        this.IsRestartRequired = sample.IsRestartRequired;
    }

    public void PrepareForReturn(ref ModelSampleStruct sample, TagRepository tagRepository)
    {
        // Copy back to unmanaged memory
        if (Values.Length > 0)
        {
            Marshal.Copy(Values, 0, sample.Values, Values.Length);
        }
        sample.ValuesCount = Values.Length;

        if (OutputValues.Length > 0)
        {
            Marshal.Copy(OutputValues, 0, sample.OutputValues, OutputValues.Length);
        }
        sample.OutputValuesCount = OutputValues.Length;

        sample.IterationIndex = Iteration;
        sample.Beta = Beta;
        sample.Z = Z;
        sample.Weight = Weight;
        sample.AllowProxy = AllowProxy;
        sample.UsedProxy = UsedProxy;
        sample.IsRestartRequired = IsRestartRequired;
        sample.Tag = tagRepository?.RegisterTag(Tag) ?? 0;
    }

    public double[] Values
    {
        get { return values; }
    }

    public double[] OutputValues
    {
        get { return outputValues; }
        set { outputValues = value; }
    }

    public int Iteration { get; set; }

    public double Beta { get; set; }

    public double Z { get; set; }

    public double Weight { get; set; }

    public bool AllowProxy { get; set; }

    public bool UsedProxy { get; set; }

    public bool IsRestartRequired { get; set; }

    public object Tag { get; set; } = null;

    public bool AreValuesEqual(ModelSample sample)
    {
        const double tolerance = 1E10;

        if (sample.Values.Length != Values.Length)
        {
            return false;
        }

        for (int i = 0; i < Values.Length; i++)
        {
            if (Math.Abs(Values[i] - sample.Values[i]) > tolerance)
            {
                return false;
            }
        }

        return true;
    }
}

[StructLayout(LayoutKind.Sequential)]
public struct ModelSampleStruct
{
    public IntPtr Values;
    public int ValuesCount;

    public IntPtr OutputValues;
    public int OutputValuesCount;

    public int IterationIndex;
    public int threadId;
    public double Weight;
    [MarshalAs(UnmanagedType.I1)] public bool AllowProxy;
    [MarshalAs(UnmanagedType.I1)] public bool UsedProxy;
    [MarshalAs(UnmanagedType.I1)] public bool IsRestartRequired;
    public double Beta;
    public double Z;
    [MarshalAs(UnmanagedType.I1)] public bool ExtendedLogging;
    public int LoggingCounter;
    public int Tag;
}
